File: backend/src/main/java/com/blockchain/EHR/controller/DoctorController.java
package com.blockchain.EHR.controller;


import com.blockchain.EHR.jwt.JwtUtils;
import com.blockchain.EHR.model.EhrDocument;
import com.blockchain.EHR.model.PatientStatus;
import com.blockchain.EHR.model.Pending;
import com.blockchain.EHR.repository.PatientRepository;
import com.blockchain.EHR.services.DoctorService;
import com.blockchain.EHR.services.EhrService;
import com.fasterxml.jackson.core.JsonProcessingException;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@CrossOrigin(origins = "*")
@Validated
@RequestMapping("/fabric/doctor")
public class DoctorController {
    @Autowired
    PatientRepository patientRepository;

    @Autowired
    private JwtUtils jwtUtils;

    @Autowired
    private DoctorService doctorService;
    @Autowired
    private com.blockchain.EHR.repository.PendingRepository pendingRepository;
    @Autowired
    private EhrService ehrService;


    @PostMapping("add-request")
    public ResponseEntity<?> addRequest(HttpServletRequest request,@RequestParam("pid")String pid){
        System.out.println("Add request");
        String jwt = jwtUtils.getJwtFromHeader(request);
        String did = jwtUtils.getUserNameFromJwtToken(jwt);
        String mspId = jwtUtils.getMspIdFromJwtToken(jwt);
        doctorService.addRequest(did,pid);
        return new ResponseEntity<>(HttpStatus.OK);
    }

    @GetMapping("/patients")
    public ResponseEntity<?> getAllPatients(HttpServletRequest request){
        String jwt = jwtUtils.getJwtFromHeader(request);
        String did = jwtUtils.getUserNameFromJwtToken(jwt);
        String mspId = jwtUtils.getMspIdFromJwtToken(jwt);

        try {
            List<PatientStatus> patientStatuses = doctorService.getPatientStatus(did,mspId);
            return new ResponseEntity<>(patientStatuses,HttpStatus.OK);
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }

    }

    // View EHR document (only if pending request status is 'Accepted')
    @GetMapping("/view-ehr")
    public ResponseEntity<EhrDocument> viewEhr(HttpServletRequest request, @RequestParam String patientId) {
        String jwt = jwtUtils.getJwtFromHeader(request);
        String did = jwtUtils.getUserNameFromJwtToken(jwt);
        String mspId = jwtUtils.getMspIdFromJwtToken(jwt);
        // Fetch the EHR document
        EhrDocument ehrDocument= ehrService.getEhrDocument(patientId,did,mspId);
        if(ehrDocument!=null){
            System.out.println("Returning document");
            return new ResponseEntity<>(ehrDocument,HttpStatus.OK);
        }
        return new ResponseEntity<>(HttpStatus.NOT_FOUND);
    }

    @GetMapping("/requests")
    public ResponseEntity<?> getAllRequests(HttpServletRequest request){
        String jwt = jwtUtils.getJwtFromHeader(request);
        String did = jwtUtils.getUserNameFromJwtToken(jwt);
        String mspId = jwtUtils.getMspIdFromJwtToken(jwt);

        List<Pending> patientStatuses = pendingRepository.findAllByDid(did);
        return new ResponseEntity<>(patientStatuses,HttpStatus.OK);
    }

    // Update EHR document (only by approved doctors)
    @PostMapping("/update-ehr")
    public ResponseEntity<String> updateEhr(HttpServletRequest request, @RequestParam String patientId,
                                            @RequestBody EhrDocument updatedEhrDocument) {
        String jwt = jwtUtils.getJwtFromHeader(request);
        String did = jwtUtils.getUserNameFromJwtToken(jwt); // Get doctor ID from JWT
        String mspId = jwtUtils.getMspIdFromJwtToken(jwt);

        // Update the EHR document
        boolean isUpdated = ehrService.updateEhr(did,patientId,mspId,updatedEhrDocument);
        if (isUpdated) {
            System.out.println("EHR document updated successfully!");
            return ResponseEntity.ok("EHR document updated successfully!");
        } else {
            System.out.println("Patient not found.");
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Patient not found.");
        }
    }

}

File: backend/src/controllers/DoctorController.ts
import { Router, Response } from 'express';
import { authMiddleware, AuthRequest } from '../middleware/authMiddleware';
import { doctorService } from '../services/DoctorService';
import { ehrService } from '../services/EhrService';
import { Pending } from '../models/Pending';

const router = Router();

/**
 * POST /fabric/doctor/add-request
 * Add access request to patient
 */
router.post('/add-request', authMiddleware, async (req: AuthRequest, res: Response) => {
    try {
        const { pid } = req.query;

        if (!pid) {
            res.status(400).json({ message: 'Patient ID is required' });
            return;
        }

        console.log('Add request');
        const did = req.user!.username;

        await doctorService.addRequest(did, pid as string);

        res.status(200).json({ message: 'Request added successfully' });
    } catch (error) {
        console.error('Add request error:', error);
        res.status(500).json({ message: 'Failed to add request' });
    }
});

/**
 * GET /fabric/doctor/patients
 * Get all accessible patients
 */
router.get('/patients', authMiddleware, async (req: AuthRequest, res: Response) => {
    try {
        const did = req.user!.username;
        const mspId = req.user!.mspId;

        const patientStatuses = await doctorService.getPatientStatus(did, mspId);

        res.status(200).json(patientStatuses);
    } catch (error) {
        console.error('Get patients error:', error);
        res.status(500).json({ message: 'Failed to get patients' });
    }
});

/**
 * GET /fabric/doctor/view-ehr
 * View patient EHR (only if access is approved)
 */
router.get('/view-ehr', authMiddleware, async (req: AuthRequest, res: Response) => {
    try {
        const { patientId } = req.query;

        if (!patientId) {
            res.status(400).json({ message: 'Patient ID is required' });
            return;
        }

        const did = req.user!.username;
        const mspId = req.user!.mspId;

        const ehrDocument = await ehrService.getEhrDocument(patientId as string, did, mspId);

        if (ehrDocument) {
            console.log('Returning EHR document');
            res.status(200).json(ehrDocument);
        } else {
            res.status(404).json({ message: 'EHR document not found or access denied' });
        }
    } catch (error) {
        console.error('View EHR error:', error);
        res.status(500).json({ message: 'Failed to view EHR' });
    }
});

/**
 * GET /fabric/doctor/requests
 * Get all requests made by doctor
 */
router.get('/requests', authMiddleware, async (req: AuthRequest, res: Response) => {
    try {
        const did = req.user!.username;

        const requests = await Pending.find({ did });

        res.status(200).json(requests);
    } catch (error) {
        console.error('Get requests error:', error);
        res.status(500).json({ message: 'Failed to get requests' });
    }
});

/**
 * POST /fabric/doctor/update-ehr
 * Update patient EHR
 */
router.post('/update-ehr', authMiddleware, async (req: AuthRequest, res: Response) => {
    try {
        const { patientId } = req.query;
        const updatedEhrDocument = req.body;

        if (!patientId) {
            res.status(400).json({ message: 'Patient ID is required' });
            return;
        }

        const did = req.user!.username;
        const mspId = req.user!.mspId;

        const isUpdated = await ehrService.updateEhr(did, patientId as string, mspId, updatedEhrDocument);

        if (isUpdated) {
            console.log('EHR document updated successfully!');
            res.status(200).send('EHR document updated successfully!');
        } else {
            console.log('Patient not found.');
            res.status(404).send('Patient not found.');
        }
    } catch (error) {
        console.error('Update EHR error:', error);
        res.status(500).json({ message: 'Failed to update EHR' });
    }
});

export default router;
